src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs

```diff
@@ -873,14 +936,68 @@ static bool MayOverlapByEnumeration(string set1, string set2)
         }
 
         /// <summary>Gets whether we can iterate through the set list pairs in order to completely enumerate the set's contents.</summary>
-        internal static bool CanEasilyEnumerateSetContents(string set) =>
+        private static bool CanEasilyEnumerateSetContents(string set) =>
             set.Length > SetStartIndex &&
             set[SetLengthIndex] > 0 &&
             set[SetLengthIndex] % 2 == 0 &&
             set[CategoryLengthIndex] == 0 &&
-            !IsNegated(set) &&
             !IsSubtraction(set);
 
+        /// <summary>Provides results from <see cref="Analyze"/>.</summary>
+        internal struct CharClassAnalysisResults
+        {
+            /// <summary>true if we know for sure that the set contains only ASCII values; otherwise, false.</summary>
+            public bool ContainsOnlyAscii;
+            /// <summary>true if we know for sure that the set doesn't contain any ASCII values; otherwise, false.</summary>
+            public bool ContainsNoAscii;
+            /// <summary>true if we know for sure that all ASCII values are in the set; otherwise, false.</summary>
+            public bool AllAsciiContained;
+            /// <summary>true if we know for sure that all non-ASCII values are in the set; otherwise, false.</summary>
+            public bool AllNonAsciiContained;
+        }
+
+        /// <summary>Analyzes the set to determine some basic properties that can be used to optimize usage.</summary>
+        internal static CharClassAnalysisResults Analyze(string set)
+        {
+            if (!CanEasilyEnumerateSetContents(set))
+            {
+                // We can't make any strong claims about the set.
+                return default;
+            }
+
+#if DEBUG
+            for (int i = SetStartIndex; i < set.Length - 1; i += 2)
+            {
+                Debug.Assert(set[i] < set[i + 1]);
+            }
+#endif
+
+            if (IsNegated(set))
+            {
+                // We're negated: if the upper bound of the range is ASCII, that means everything
+                // above it is actually included, meaning all non-ASCII are in the class.
+                // Similarly if the lower bound is non-ASCII, that means in a negated world
+                // everything ASCII is included.
+                return new CharClassAnalysisResults
+                {
+                    AllNonAsciiContained = set[^1] < 128,
+                    AllAsciiContained = set[SetStartIndex] >= 128,
+                    ContainsNoAscii = false,
```

Comment:
What if the set was `[^\u0000-\u007F]` couldn't we say ContainsNoAscii ==true?

Reply:
We could.  We'd need to check for that exact range.  Not sure it's worth it.